example/eblob_backend.c

```diff
@@ -872,13 +874,156 @@ int blob_defrag_start(void *priv, enum dnet_backend_defrag_level level)
+
+	/*
+	 * Set NEED_ACK bit to signal server-send controller that we want
+	 * to send final ACK when controller will be destroyed, which in turn
+	 * will happend after all WRITE commands are completed.
+	 *
+	 * Command will be copied internally in @dnet_server_send_alloc(),
+	 * thus it is safe to clear that bit afterward.
+	 */
+	cmd->flags |= DNET_FLAGS_NEED_ACK;
+
+	ctl = dnet_server_send_alloc(state, cmd, req->iflags, groups, req->group_num);
+	if (!ctl) {
+		err = -ENOMEM;
+		goto err_out_exit;
+	}
+
+	/*
+	 * Deliberately clear NEED_ACK bit
+	 * This function will iterate over provided ids,
+	 * read them from the blob and queue WRITE command to remote groups.
+	 * When single WRITE command completes, it will send response back to client.
+	 *
+	 * If we send ACK in the middle, it will force client to stop accepting further responses.
+	 *
+	 * If there will be an error, it will be sent to client too.
+	 * If there is an error with command processing (like fail to allocate memory),
+	 * iteration will stop and error will be returned from this function to the higher layer,
+	 * which in turn will force ACK message to client with error code.
+	 */
+	cmd->flags &= ~DNET_FLAGS_NEED_ACK;
+
+
+
+	for (i = 0; i < req->id_num; ++i) {
+		memcpy(key.id, ids[i].id, EBLOB_ID_SIZE);
+
+		err = blob_lookup(b, &key, &wc);
+		if (err < 0) {
+			dnet_backend_log(cfg->blog, DNET_LOG_ERROR, "%s: EBLOB: blob_send: lookup: %d: %s",
+					 dnet_dump_id_str(key.id), err, strerror(-err));
+			goto err_out_send_fail_reply;
+		}
+
+		re.key = ids[i];
+		re.flags = wc.flags; // these flags corresponse to DNET_RECORD_FLAGS_*
+		re.status = 0;
+		re.iterated_keys = i;
+		re.size = wc.total_data_size;
+		// set iterator response id to differentiate various commands
+		// client can use cmd->backend_id from reply though
+		re.id = cmd->backend_id;
+
+		data_offset = wc.data_offset;
+		record_offset = 0;
+
+		if ((wc.flags & BLOB_DISK_CTL_EXTHDR) != 0) {
+			struct dnet_ext_list_hdr ehdr;
+
+			/* Sanity */
+			if (re.size < ehdr_size) {
+				err = -ERANGE;
+				goto err_out_send_fail_reply;
+			}
+
+			err = dnet_ext_hdr_read(&ehdr, wc.data_fd, wc.data_offset);
+			if (err != 0)
+				goto err_out_send_fail_reply;
+
+			dnet_ext_hdr_to_list(&ehdr, &elist);
+
+			re.timestamp = elist.timestamp;
+			re.user_flags = elist.flags;
+
+			/* Take into an account extended header's len */
+			re.size -= sizeof(struct dnet_ext_list_hdr);
+			data_offset += sizeof(struct dnet_ext_list_hdr);
+			record_offset += sizeof(struct dnet_ext_list_hdr);
+		}
+
+		wc.offset = record_offset;
+		wc.size = re.size;
+		err = eblob_verify_checksum(b, &key, &wc);
+		if (err)
+			goto err_out_send_fail_reply;
+
+		err = dnet_server_send_write(ctl, &re, sizeof(struct dnet_iterator_response), wc.data_fd, data_offset);
+		if (err)
+			goto err_out_send_fail_reply;
+
+		continue;
+
+err_out_send_fail_reply:
+		re.status = err;
+		err = blob_send_reply(state, cmd, &re, 1);
```

Comment:
Will it really skip all keys after it faces the first one which:
* doesn't exist
* or has broken ext header
* or has broken checksum
* or it has `dnet_server_send_write` failed